metricflow/plan_conversion/dataflow_to_sql.py

```diff
@@ -1297,3 +1301,64 @@ def visit_semi_additive_join_node(self, node: SemiAdditiveJoinNode) -> SqlDataSe
                 order_bys=(),
             ),
         )
+
+    def visit_join_to_time_spine_node(self, node: JoinToTimeSpineNode[SourceDataSetT]) -> SqlDataSet:  # noqa: D
+        parent_data_set = node.parent_node.accept(self)
+        parent_alias = self._next_unique_table_alias()
+
+        # Build time spine dataset
+        metric_time_dimension_instance: Optional[TimeDimensionInstance] = None
+        for instance in parent_data_set.metric_time_dimension_instances:
+            if len(instance.spec.identifier_links) == 0:
+                metric_time_dimension_instance = instance
+                break
+        assert (
+            metric_time_dimension_instance
+        ), "Can't query offset metric without a time dimension. Validations should have prevented this."
+        metric_time_dimension_column_name = self.column_association_resolver.resolve_time_dimension_spec(
+            metric_time_dimension_instance.spec
+        ).column_name
+        time_spine_alias = self._next_unique_table_alias()
+        time_spine_dataset = _make_time_spine_data_set(
+            metric_time_dimension_instance=metric_time_dimension_instance,
+            metric_time_dimension_column_name=metric_time_dimension_column_name,
+            time_spine_source=self._time_spine_source,
+            time_spine_table_alias=time_spine_alias,
+            time_range_constraint=node.time_range_constraint,
+        )
+
+        # Build join expression
+        left_expr: SqlExpressionNode = SqlColumnReferenceExpression(
+            col_ref=SqlColumnReference(table_alias=time_spine_alias, column_name=metric_time_dimension_column_name)
+        )
+        if node.offset_window:
+            left_expr = SqlTimeDeltaExpression(
+                arg=left_expr, count=node.offset_window.count, granularity=node.offset_window.granularity
+            )
```

Comment:
Ah, ok, I see. The offset here is always exact to granularity, not capturing a range, right? I mean if it's 10 days, we don't look for all dates between 10 days ago, we match on things that were exactly 10 days ago.

Note this is applied to the time spine, not the metric data set. So we're offsetting the spine and filtering out any rows from 10 days after the start of the time spine.

Reply:
Correct on both counts!

> So we're offsetting the spine and filtering out any rows from 10 days after the start of the time spine.
Hmmm. Do we worry that users will want to include those missing 10 days that get removed from the time spine? When querying an offset metric, we could adjust the requested time range constraint when there is an offset window. But that won't fix the issue if they're querying from the start of the time range available in the time spine.

Reply:
I *think* this gets resolved when we grab the time_spine metric_time. The only missing values in that case would be values that are missing in the data.

It'd be good to have our integration test cases in the next PR check that boundary so we can see where the NULLs show up, but I'm pretty sure your next update will address this!